src/service/layers-impl.js

```diff
@@ -0,0 +1,634 @@
+    this.needsScrollRemeasure_ = true;
+
+    const element = this.element_;
+    const scroller = getScrollingElement(element);
+    this.isRootLayer_ = element === scroller || !scroller.contains(element);
+
+    const parent = this.getParentLayer();
+    if (parent) {
+      parent.transfer_(this);
+    }
+  }
+
+  /**
+   * This is necessary when the screen size changes (or other similar events
+   * happen), because CSS styles may change. We'll need to recompute layers
+   * lazily after this happens.
+   */
+  undeclareLayer() {
+    if (!this.isLayer_ || this.isRootLayer_) {
+      return;
+    }
+
+    this.isLayer_ = false;
+    this.transfer_(this.getParentLayer());
+  }
+
+  transfer_(layer) {
+    // An optimization if we know that the new layer definitely contains
+    // everything in this layer.
+    const contained = layer.contains(this.element_);
+
+    filterSplice(this.children_, layout => {
+      if (contained || layer.contains(layout.element_)) {
+        layout.parentLayer_ = layer;
+        layer.children_.push(layout);
+        return false;
+      }
+
+      return true;
+    });
+  }
+
+  /**
+   * Gets the element's parent layer. If this element is itself a layer, it
+   * returns the layer's parent.
+   * @return {?LayoutElement}
+   */
+  getParentLayer() {
+    if (this.parentLayer_ === undefined) {
+      const parent = LayoutElement.getParentLayer(this.element_, true);
+      this.parentLayer_ = parent;
+      if (parent) {
+        parent.add(this.element_);
+      }
+    }
+    return this.parentLayer_;
+  }
+
+  forgetParentLayer() {
+    this.parentLayer_ = undefined;
+  }
+
+  /**
+   * Gets the size of the element.
+   * @return {!SizeDef}
+   */
+  getSize() {
+    if (this.needsRemeasure_) {
+      this.remeasure();
+    }
+    return this.size_;
+  }
+
+  /**
+   * Gets offsets of the element relative to the parent layer, without taking
+   * scroll position into account.
+   * @return {!PositionDef}
+   */
+  getOffsetFromParent() {
+    if (this.needsRemeasure_) {
+      this.remeasure();
+    }
+    return this.position_;
+  }
+
+  getScrollTop() {
+    this.updateScrollPosition_();
+    return this.scrollTop_;
+  }
+
+  getScrollLeft() {
+    this.updateScrollPosition_();
+    return this.scrollLeft_;
+  }
+
+  /**
+   * Gets the position of the element relative to the parent layer, taking
+   * scroll position of the parents into account.
+   * @param {Element=} opt_parent
+   * @return {!LayoutRectDef}
+   */
+  getScrolledPosition(opt_parent) {
+    const position = this.getOffsetFromParent();
+    let x = position.left;
+    let y = position.top;
+
+    let last;
+    const stopAt = opt_parent ? LayoutElement.getParentLayer(opt_parent) : null;
+    for (let p = this.getParentLayer(); p !== stopAt; p = p.getParentLayer()) {
+
+      x -= p.getScrollLeft();
+      y -= p.getScrollTop();
+
+      if (last) {
+        const position = last.getOffsetFromParent();
+        x += position.left;
```

Comment:
Overall, I'm always suspicious of plus/minus logic in with elements nested in a complicated way. The deeper is the nesting, the harder it is assign any real meaning to such numbers. The most clear interpretation are intersection APIs (e.g.  `intersection(withLayerOrViewport):number`). And scoring numbers are obviously also meaningful.

Reply:
I'm not sure how to address this comment. The scrolled position is a core necessity of Layers, and can't be changed. Would comments throughout (which I need to do as part of the documentation anyways) help?

Reply:
Internally we can do math as we need. My question is what API we export outside and what API is actually needed. Local position, i.e. `{layer, topInLayer, leftInLayer}` totally always makes sense. But absolute position might not really exist in a general case, or it could be some weird that it'd be only usable in few cases, such as document ad position. But for runtime's scheduling what do we need?

- In-layer local top/left: clear.
- Intersection with any layer/viewport: intersection ratio is a great API and very clear and unambiguous.
- Scheduling scores: might be easier to pass a hierarchy-walker style callback to calculate double values?

Also, I have a performance question about the same: in theory calculating intersection ratio and distance scores does not require us to constantly recalculate structs of the form `{top, left, right, bottom}`. This likely would be impactful if these structs depend on scroll position and would help reduce object creates.

Reply:
Ah, for external (non-Layers) APIs, I've gone through a few thoughts. We can either move everything into Layers (a topic I discussed with Will) which seems like an acceptable first solution.

But the best solution I can think of at the moment is to provide a callback API, similar to `#forEach`. But I do **not** want to expose the Layer/Elements to the callback. My current thought is that we'll pass `top, left, containerWidth, containerHeight` as parameters:

```js
// Just calling it forEach for now
layout.forEach((top, left, containerWidth, containerHeight) => {
  // top is actually the scrolled top, ie, offset.top - container.scrollTop
  // same for left.

  // Calculate intersection
  // Or, calculate scoring
});
```

At the moment, these seem to be the only values necessary for non-Layers code. This also nicely skirts the issue of structs, since the values can be calculated without creating any temporary objects (offsets/scrollPositions/sizes will already be cached on each Element), and anything that will change frequently (scroll position) is stored as two numbers directly on the Element instead of a constantly invalid struct.

Reply:
Yeah. This feels right. The list of args may get a bit unwieldy, but if you feel strongly about hiding layer structure - sure.